atef/widgets/config/data_active.py

```diff
@@ -927,13 +931,37 @@ def update_target(self) -> None:
 
         self.update_input_placeholder()
 
+    def get_curr_value(self):
+        if self._sig is None:
+            return
+        self._sig.wait_for_connection()
+        self._curr_value = self.bridge.value.get() or self._sig.get()
+        self._dtype = type(self._curr_value)
+        self._enum_strs = getattr(self._sig, 'enum_strs', None)
+
+    def fail_get_value(self, ex: Exception):
+        logger.debug(f'failed to get signal data for input widget: {ex}')
+        self._curr_value = 'no data'
+        # fall back to type in dataclass if available
+        stored_value = self.bridge.value.get()
+        if stored_value is not None:
+            self._dtype = type(stored_value)
+        else:
+            self._dtype = float
+
+        self.run_setup_input_widget()
+
+    def run_setup_input_widget(self):
+        self.setup_input_widget(self._curr_value, self._dtype,
+                                enum_strs=self._enum_strs)
+
     def update_input_placeholder(self) -> None:
         """
         Updates value input widget with a QLineEdit with the approriate validator
         given the target's datatype
         """
-        sig: EpicsSignalBase = self.data.to_signal()
-        if sig is None:
+        self._sig: EpicsSignalBase = self.data.to_signal()
+        if self._sig is None:
```

Comment:
This bit is confusing from a type hinting perspective and probably gives some of the stricter type checking libraries pause. It's saying "if this thing that is always an EpicsSignalBase is None..."

Maybe this type should be `EpicsSignalBase | None` and maybe it should be hinted in the object init or in the class definition itself instead of in this method?

Reply:
It might also just be a regular signal, right?

Reply:
Turns out `atef.Procedure.Target.to_signal() -> Optional[EpicsSignal]` as written currently, so I'll keep it as such and move it to the init